pennylane/devices/qubit_mixed/apply_operation.py

```diff
@@ -672,3 +696,158 @@ def apply_snapshot(
             debugger.snapshots[len(debugger.snapshots)] = snapshot
 
     return state
+
+
+# pylint: disable=unused-argument
+@apply_operation.register
+def apply_density_matrix(
+    op: qml.QubitDensityMatrix,
+    state,
+    is_state_batched: bool = False,
+    debugger=None,
+    **execution_kwargs,
+):
+    """
+    Applies a QubitDensityMatrix operation by initializing or replacing
+    the quantum state with the provided density matrix.
+
+    - If the QubitDensityMatrix covers all wires, we directly return the provided density matrix as the new state.
+    - If only a subset of the wires is covered, we:
+      1. Partial trace out those wires from the current state to get the density matrix of the complement wires.
+      2. Take the tensor product of the complement density matrix and the provided density_matrix.
+      3. Reshape to the correct final shape and return.
+
+    Args:
+        op (qml.QubitDensityMatrix): The QubitDensityMatrix operation.
+        state (array-like): The current quantum state.
+        is_state_batched (bool): Whether the state is batched.
+        debugger: A debugger instance for diagnostics.
+        **execution_kwargs: Additional kwargs.
+
+    Returns:
+        array-like: The updated quantum state.
+
+    Raises:
+        ValueError: If the density matrix is invalid.
+    """
+    density_matrix = op.parameters[0]
+    num_wires = len(op.wires)
+    expected_dim = 2**num_wires
+
+    # Cast density_matrix to the same type and device as state
+    density_matrix = math.cast_like(density_matrix, state)
+
+    # Extract total wires
+    num_state_wires = _get_num_wires(state, is_state_batched)
+    all_wires = list(range(num_state_wires))
+    op_wires = op.wires
+    complement_wires = [w for w in all_wires if w not in op_wires]
+
+    # If the operation covers the full system, just return it
+    if len(op_wires) == num_state_wires:
+        # If batched, broadcast
+        if is_state_batched:
+            batch_size = math.shape(state)[0]
+            density_matrix = math.broadcast_to(
+                density_matrix, (batch_size,) + math.shape(density_matrix)
+            )
+
+        # Reshape to match final shape of state
+        return math.reshape(density_matrix, math.shape(state))
+
+    # Partial system update:
+    # 1. Partial trace out op_wires from state
+    # partial_trace reduces the dimension to only the complement wires
+    sigma = qml.math.partial_trace(state, indices=op_wires)
+    # sigma now has shape:
+    # (batch_size?, 2^(n - num_wires), 2^(n - num_wires)) where n = total wires
```

Comment:
```suggestion
    # (batch_size?, 2^(n - num_wires), 2^(n - num_wires)) where n = total wires
```

What does the question mark next to `batch_size` mean?

Reply:
Will del `?` in the batch apply suggestions